design/ceph/ceph-orchestrator-osd-add-remove.md

```diff
@@ -0,0 +1,147 @@
+not keep track of the Drive Groups specification used.
+
+Ceph expects the on-disk content of disks to be the declarative state of whether an OSD belongs to a
+Ceph cluster or not and what its configuration is. Rook is currently able to read this on-disk
+content via `ceph-volume` to determine if it should run a Kubernetes Deployment for a disk-based
+OSD.
+
+Rook currently does not implement OSD removal via any declarations in the `CephCluster` resource in
+order to prioritize the safety of user data. If a user wishes to remove an OSD, they must perform
+manual, imperative steps to do so.
+
+## Proposal
+
+The chief difficulty with this proposal is to harmonize interactions between Rook, which ideally
+wants to use Kubernetes manifests for a declarative intent for a Ceph cluster, and the Ceph
+Orchestrator Mgr Module, which treats OSD creation as imperative and allows the on-disk content of
+OSDs to be the declarative state.
+
+### Adding OSDs
+This design proposes to add a section to the `CephCluster` resource's `storage` configuration
+section for individual `nodes` to have disks added via a Drive Groups specification. The spec is
+proposed below with behavior described to follow.
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephCluster
+spec:
+  storage:
+    nodes: # preexisting
+      - name: < node name > # preexisting
+        driveGroups: # proposed new
+          - removeFromSpecAfterCompmletion: < default false > # proposed new
+            spec: < JSON string to pass to ceph-volume for OSD creation >
+```
+
+Each individual `driveGroups` item's `spec` will be a simple string containing a JSON blob that gets
+passed along without modification or inspection by Rook to `ceph-volume` when running the OSD
+provisioning job on the corresponding node. The `ceph-volume` command to create OSDs via the Drive
+Groups spec is given priority and executed first before any other Rook-based OSD creation methods
+are used.
+
+Users using `CephCluster` via Kubernetes Manifest will now be able to supply their own Drive Groups
+spec to Rook if they so desire to allow themselves greater flexibility in OSD management. Rook will
+treat this user spec as a declarative intent for the cluster so long as the
+`removeFromSpecAfterCompletion` is unspecified or set to `false`. Setting this value to `true` will
+change Rook's behavior to treat the spec imperatively, intended for use by the Orchestrator Mgr
+Module. Upon completion of the `ceph-volume` command with the Drive Group spec, Rook will remove the
+Drive Group item from the `driveGroups` spec for the corresponding node in the `CephCluster`
+resource so that Rook will not rerun the creation again, which would break the Ceph Orchestrator Mgr
+Module's imperative intention. The Mgr Module should obviously set this value to `true` for all
+operations.
+
+**TODO DISCUSS THIS:**
+The Drive Group item should be removed regardless of whether the `ceph-volume` command result was
```

Comment:
Instead of removing the driveGroup from the CR automatically, how about if we use the status as you suggest, but allow the mgr module to make the decision to remove it? The mgr module could remove it if it succeeded. This also gives a feedback mechanism where the mgr module could report back to the dashboard that the driveGroup configuration failed and the user could choose to retry (we would re-trigger an orchestration with the same CR), or abort (remove the driveGroup from the CR).

Reply:
I think this is possible, but it does worry me somewhat. I want to discuss in the orchestrator Monday meeting about the way the orchestrator expects to kick off actions and how it expects to get results back.

I worry somewhat about what happens if the mgr goes offline in some sort of longer term failure mode after adding a Drive Group to Rook. Rook might continue to execute the Drive Group while the mgr is offline. Although the user is hopefully not adding disks to nodes while the mgr is offline, it is possible. This is the case where I think it might be worth considering Rook removing things from the CRD spec as an exception here.

If we absolutely will not allow rook to remove CRD spec items, I wonder if it would be a good idea to include in the `status` spec for Drive Groups a count of the number of times Rook has attempted to run `ceph-volume` with the spec. This might be a good idea, even if Rook is removing CRD items in the case where the operator crashes between adding the status item and removing the spec item.

Reply:
After some more reading on custom resources, I find that (1) the resource controller should not alter the `spec`, and (2) the user should not alter the `status`. Removing items from the spec as a user does seem dangerous, as there could be race conditions between Rook adding a condition and the user removing one, though I don't think this would be a terrible issue, but if we can avoid it that would be best.

Maybe a way to do what we need while still adhering to the above guidelines would be to have the operator create a status for each Drive Group spec item. If the user (or mgr in this case) removes a Drive Group spec item, the operator would remove the item from the status as well.

Reply:
I still see that Rook CRs should be desired state and from rook's perspective will always be that way. If the mgr module needs to tweak the CR so Rook stops re-applying it, let's discuss from that perspective.

Another consideration is upgrades. When Rook needs to upgrade the OSDs it needs to re-discover them from the osd-prepare pod so the operator will know that the daemons exist and should be updated. I'm assuming we can still rely on `ceph-volume lvm list` (or similar command) as we already use today to discover the existing OSDs. The key is that we at least need to trigger the osd prepare job on the nodes where driveGroups have been configured. If the driveGroups are removed from the CR, we have to make sure the operator still knows to check those nodes for OSDs.

The `status` section is not editable by the user, so they wouldn't be able to remove it if they want to retry.

Reply:
@travisn this is at least hinted at in the section discussing what is out of scope. Ceph Orchestrator expects to add nodes to the cluster via `node add` commands. Added nodes should appear in the `spec:storage:nodes` section of the CRD, and I think it is fine if Rook doesn't look for new OSDs on nodes that are not in the CRD. However, if there are OSD deployments on a node, even if that node is not in the `spec:storage:nodes` section, I think Rook should still check that node. This is out of scope here I think, but it might make sense to formalize the node add/rm functionality in a separate design PR if you think that would be necessary.
@sebastian-philipp might also find that useful?